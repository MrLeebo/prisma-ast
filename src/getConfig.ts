import type { IParserConfig } from 'chevrotain';
import {
  lilconfigSync as configSync,
  type LilconfigResult as ConfigResultRaw,
} from 'lilconfig';

export type PrismaAstParserConfig = Pick<IParserConfig, 'nodeLocationTracking'>;
export interface PrismaAstConfig {
  parser: PrismaAstParserConfig;
}

type ConfigResult<T> = Omit<ConfigResultRaw, 'config'> & {
  config: T;
};

const defaultConfig: PrismaAstConfig = {
  parser: { nodeLocationTracking: 'none' },
};

let config: PrismaAstConfig;
export default function getConfig(): PrismaAstConfig {
  if (config != null) return config;

  const result: ConfigResult<PrismaAstConfig> | null =
    configSync('prisma-ast').search();
  return (config = Object.assign(defaultConfig, result?.config));
}
