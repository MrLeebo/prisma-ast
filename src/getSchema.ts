import { PrismaLexer } from './lexer';
import { PrismaVisitor } from './visitor';
import { parser } from './parser';

export function getSchema(source: string): Schema {
  const lexingResult = PrismaLexer.tokenize(source);
  parser.input = lexingResult.tokens;
  const cstNode = parser.schema();
  if (parser.errors.length > 0) throw parser.errors[0];
  const visitor = new PrismaVisitor();
  return visitor.visit(cstNode);
}

export interface Schema {
  type: 'schema';
  list: Block[];
}

export type Block =
  | Model
  | View
  | Datasource
  | Generator
  | Enum
  | Comment
  | Break;

export interface Object {
  type: 'model' | 'view';
  name: string;
  properties: Array<Property | Comment | Break>;
}

export interface Model extends Object {
  type: 'model';
}

export interface View extends Object {
  type: 'view';
}

export interface Datasource {
  type: 'datasource';
  name: string;
  assignments: Array<Assignment | Comment | Break>;
}

export interface Generator {
  type: 'generator';
  name: string;
  assignments: Array<Assignment | Comment | Break>;
}

export interface Enum {
  type: 'enum';
  name: string;
  enumerators: Array<Enumerator | Comment | Break>;
}

export interface Comment {
  type: 'comment';
  text: string;
}

export interface Break {
  type: 'break';
}

export type Property = GroupedBlockAttribute | BlockAttribute | Field;

export interface Assignment {
  type: 'assignment';
  key: string;
  value: Value;
}

export interface Enumerator {
  type: 'enumerator';
  name: string;
  value?: Value;
  comment?: string;
}

export interface BlockAttribute {
  type: 'attribute';
  kind: 'object' | 'view';
  group?: string;
  name: string;
  args: AttributeArgument[];
}

export type GroupedBlockAttribute = BlockAttribute & { group: string };

export interface Field {
  type: 'field';
  name: string;
  fieldType: string | Func;
  array?: boolean;
  optional?: boolean;
  attributes?: Attribute[];
  comment?: string;
}

export type Attr =
  | Attribute
  | GroupedAttribute
  | BlockAttribute
  | GroupedBlockAttribute;

export interface Attribute {
  type: 'attribute';
  kind: 'field';
  group?: string;
  name: string;
  args?: AttributeArgument[];
}

export type GroupedAttribute = Attribute & { group: string };

export interface AttributeArgument {
  type: 'attributeArgument';
  value: KeyValue | Value | Func;
}

export interface KeyValue {
  type: 'keyValue';
  key: string;
  value: Value;
}

export interface Func {
  type: 'function';
  name: string;
  params: Value[];
}

export interface RelationArray {
  type: 'array';
  args: string[];
}

export type Value =
  | string
  | number
  | boolean
  | Func
  | RelationArray
  | Array<Value>;
