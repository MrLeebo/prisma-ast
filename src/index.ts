export * from './produceSchema';
export * from './getSchema';
export * from './printSchema';
export * from './PrismaSchemaBuilder';
export type { PrismaAstConfig } from './getConfig';
export type { CstNodeLocation } from 'chevrotain';
export { VisitorClassFactory } from './visitor';
export { PrismaParser } from './parser';
