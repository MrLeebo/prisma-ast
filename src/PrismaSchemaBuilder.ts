import * as schema from './getSchema';
import { isSchemaField, isSchemaObject } from './schemaUtils';
import { PrintOptions, printSchema } from './printSchema';

/** Returns the function type Original with its return type changed to NewReturn. */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ReplaceReturnType<Original extends (...args: any) => any, NewReturn> = (
  ...a: Parameters<Original>
) => NewReturn;

/**
 * Methods with return values that do not propagate the builder should not have
 * their return value modified by the type replacement system below
 * */
type ExtractKeys = 'getSchema' | 'getSubject' | 'getParent' | 'print';

/** These keys preserve the return value context that they were given */
type NeutralKeys = 'break' | 'comment' | 'attribute' | 'enumerator' | 'then';

/** Keys allowed when you call .datasource() or .generator() */
type DatasourceOrGeneratorKeys = 'assignment';

/** Keys allowed when you call .enum("name") */
type EnumKeys = 'enumerator';

/** Keys allowed when you call .field("name") */
type FieldKeys = 'attribute' | 'removeAttribute';

/** Keys allowed when you call .model("name") */
type BlockKeys = 'blockAttribute' | 'field' | 'removeField';

/**
 * Utility type for making the PrismaSchemaBuilder below readable:
 * Removes methods from the builder that are prohibited based on the context
 * the builder is in. For example, you can add fields to a model, but you can't
 * add fields to an enum or a datasource.
 */
type PrismaSchemaSubset<
  Universe extends keyof ConcretePrismaSchemaBuilder,
  Method
> = ReplaceReturnType<
  ConcretePrismaSchemaBuilder[Universe],
  PrismaSchemaBuilder<Exclude<keyof ConcretePrismaSchemaBuilder, Method>>
>;

/**
 * The brain of this whole operation: depending on the key of the method name
 * we receive, filter the available list of method calls the user can make to
 * prevent them from making invalid calls, such as builder.datasource().field()
 * */
type PrismaSchemaBuilder<K extends keyof ConcretePrismaSchemaBuilder> = {
  [U in K]: U extends ExtractKeys
    ? ConcretePrismaSchemaBuilder[U]
    : U extends NeutralKeys
    ? ConcretePrismaSchemaBuilder[U] //ReplaceReturnType<ConcretePrismaSchemaBuilder[U], PrismaSchemaBuilder<K>>
    : U extends 'datasource'
    ? PrismaSchemaSubset<U, 'datasource' | EnumKeys | FieldKeys | BlockKeys>
    : U extends 'generator'
    ? PrismaSchemaSubset<U, EnumKeys | FieldKeys | BlockKeys>
    : U extends 'model'
    ? PrismaSchemaSubset<U, DatasourceOrGeneratorKeys | EnumKeys | FieldKeys>
    : U extends 'view'
    ? PrismaSchemaSubset<U, DatasourceOrGeneratorKeys | EnumKeys | FieldKeys>
    : U extends 'field'
    ? PrismaSchemaSubset<U, DatasourceOrGeneratorKeys | EnumKeys>
    : U extends 'removeField'
    ? PrismaSchemaSubset<U, DatasourceOrGeneratorKeys | EnumKeys | FieldKeys>
    : U extends 'enum'
    ? PrismaSchemaSubset<U, DatasourceOrGeneratorKeys | BlockKeys | FieldKeys>
    : U extends 'removeAttribute'
    ? PrismaSchemaSubset<U, DatasourceOrGeneratorKeys | EnumKeys>
    : PrismaSchemaSubset<
        U,
        DatasourceOrGeneratorKeys | EnumKeys | FieldKeys | BlockKeys | 'comment'
      >;
};

type Arg =
  | string
  | {
      name: string;
      function?: Arg[];
    };
type Parent = schema.Block | undefined;
type Subject = schema.Block | schema.Field | undefined;

export class ConcretePrismaSchemaBuilder {
  private schema: schema.Schema;
  private _subject: Subject;
  private _parent: Parent;

  constructor(source = '') {
    this.schema = schema.getSchema(source);
  }

  /** Prints the schema out as a source string */
  print(options: PrintOptions = {}): string {
    return printSchema(this.schema, options);
  }

  /** Returns the underlying schema object for more advanced use cases. */
  getSchema(): schema.Schema {
    return this.schema;
  }

  /** Adds or updates a generator block based on the name. */
  generator(name: string, provider = 'prisma-client-js'): this {
    const generator: schema.Generator = this.schema.list.reduce<
      schema.Generator
    >(
      (memo, block) =>
        block.type === 'generator' && block.name === name ? block : memo,
      {
        type: 'generator',
        name,
        assignments: [
          { type: 'assignment', key: 'provider', value: `"${provider}"` },
        ],
      }
    );

    if (!this.schema.list.includes(generator)) this.schema.list.push(generator);
    this._subject = generator;
    return this;
  }

  drop(name: string): this {
    const index = this.schema.list.findIndex(
      block => 'name' in block && block.name === name
    );
    this.schema.list.splice(index, 1);
    return this;
  }

  /** Sets the datasource for the schema. */
  datasource(provider: string, url: string | { env: string }): this {
    const datasource: schema.Datasource = {
      type: 'datasource',
      name: 'db',
      assignments: [
        {
          type: 'assignment',
          key: 'url',
          value:
            typeof url === 'string'
              ? `"${url}"`
              : { type: 'function', name: 'env', params: [`"${url.env}"`] },
        },
        { type: 'assignment', key: 'provider', value: provider },
      ],
    };
    const existingIndex = this.schema.list.findIndex(
      block => block.type === 'datasource'
    );
    this.schema.list.splice(
      existingIndex,
      existingIndex !== -1 ? 1 : 0,
      datasource
    );
    this._subject = datasource;
    return this;
  }

  /** Adds or updates a model based on the name. Can be chained with .field() or .blockAttribute() to add to it. */
  model(name: string): this {
    const model = this.schema.list.reduce<schema.Model>(
      (memo, block) =>
        block.type === 'model' && block.name === name ? block : memo,
      { type: 'model', name, properties: [] }
    );
    if (!this.schema.list.includes(model)) this.schema.list.push(model);
    this._subject = model;
    return this;
  }

  /** Adds or updates a view based on the name. Can be chained with .field() or .blockAttribute() to add to it. */
  view(name: string): this {
    const view = this.schema.list.reduce<schema.View>(
      (memo, block) =>
        block.type === 'view' && block.name === name ? block : memo,
      { type: 'view', name, properties: [] }
    );
    if (!this.schema.list.includes(view)) this.schema.list.push(view);
    this._subject = view;
    return this;
  }

  /** Adds or updates an enum based on the name. Can be chained with .enumerator() to add a value to it. */
  enum(name: string, enumeratorNames: string[] = []): this {
    const e = this.schema.list.reduce<schema.Enum>(
      (memo, block) =>
        block.type === 'enum' && block.name === name ? block : memo,
      {
        type: 'enum',
        name,
        enumerators: enumeratorNames.map(name => ({
          type: 'enumerator',
          name,
        })),
      }
    );
    if (!this.schema.list.includes(e)) this.schema.list.push(e);
    this._subject = e;
    return this;
  }

  enumerator(value: string): this {
    const subject = this.getSubject<schema.Enum>();
    if (!subject || !('type' in subject) || subject.type !== 'enum') {
      throw new Error('Subject must be a prisma enum!');
    }

    subject.enumerators.push({ type: 'enumerator', name: value });
    return this;
  }

  /**
   * Returns the current subject, such as a model, field, or enum.
   * @example
   * builder.getModel('User').field('firstName').getSubject() // the firstName field
   * */
  private getSubject<S extends Subject>(): S {
    return this._subject as S;
  }

  /** Returns the parent of the current subject when in a nested context. The parent of a field is its model or view. */
  private getParent<S extends Parent = schema.Object>(): S {
    return this._parent as S;
  }

  /**
   * Adds a block-level attribute to the current model.
   * @example
   * builder.model('Project')
   *   .blockAttribute("map", "projects")
   *   .blockAttribute("unique", ["firstName", "lastName"]) // @@unique([firstName, lastName])
   * */
  blockAttribute(
    name: string,
    args?: string | string[] | Record<string, schema.Value>
  ): this {
    let subject = this.getSubject<schema.Object>();
    if (!isSchemaObject(subject)) {
      const parent = this.getParent<schema.Object>();
      if (!isSchemaObject(parent))
        throw new Error('Subject must be a prisma model or view!');

      subject = this._subject = parent;
    }

    const attributeArgs = ((): schema.AttributeArgument[] => {
      if (!args) return [] as schema.AttributeArgument[];
      if (typeof args === 'string')
        return [{ type: 'attributeArgument', value: `"${args}"` }];
      if (Array.isArray(args))
        return [{ type: 'attributeArgument', value: { type: 'array', args } }];
      return Object.entries(args).map(([key, value]) => ({
        type: 'attributeArgument',
        value: { type: 'keyValue', key, value },
      }));
    })();

    const property: schema.BlockAttribute = {
      type: 'attribute',
      kind: 'object',
      name,
      args: attributeArgs,
    };
    subject.properties.push(property);
    return this;
  }

  /** Adds an attribute to the current field. */
  attribute<T extends schema.Field>(
    name: string,
    args?: Arg[] | Record<string, string[]>
  ): this {
    const parent = this.getParent();
    const subject = this.getSubject<T>();
    if (!isSchemaObject(parent)) {
      throw new Error('Parent must be a prisma model or view!');
    }

    if (!isSchemaField(subject)) {
      throw new Error('Subject must be a prisma field!');
    }

    if (!subject.attributes) subject.attributes = [];
    const attribute = subject.attributes.reduce<schema.Attribute>(
      (memo, attr) =>
        attr.type === 'attribute' &&
        `${attr.group ? `${attr.group}.` : ''}${attr.name}` === name
          ? attr
          : memo,
      {
        type: 'attribute',
        kind: 'field',
        name,
      }
    );

    if (Array.isArray(args)) {
      const mapArg = (arg: Arg): schema.Value | schema.Func => {
        return typeof arg === 'string'
          ? arg
          : {
              type: 'function',
              name: arg.name,
              params: arg.function?.map(mapArg) ?? [],
            };
      };

      if (args.length > 0)
        attribute.args = args.map(arg => ({
          type: 'attributeArgument',
          value: mapArg(arg),
        }));
    } else if (typeof args === 'object') {
      attribute.args = Object.entries(args).map(([key, value]) => ({
        type: 'attributeArgument',
        value: { type: 'keyValue', key, value: { type: 'array', args: value } },
      }));
    }

    if (!subject.attributes.includes(attribute))
      subject.attributes.push(attribute);

    return this;
  }

  /** Remove an attribute from the current field */
  removeAttribute<T extends schema.Field>(name: string): this {
    const parent = this.getParent();
    const subject = this.getSubject<T>();
    if (!isSchemaObject(parent)) {
      throw new Error('Parent must be a prisma model or view!');
    }

    if (!isSchemaField(subject)) {
      throw new Error('Subject must be a prisma field!');
    }

    if (!subject.attributes) subject.attributes = [];
    subject.attributes = subject.attributes.filter(
      attr => !(attr.type === 'attribute' && attr.name === name)
    );

    return this;
  }

  /** Add an assignment to a generator or datasource */
  assignment<T extends schema.Generator | schema.Datasource>(
    key: string,
    value: string
  ): this {
    const subject = this.getSubject<T>();
    if (
      !subject ||
      !('type' in subject) ||
      !['generator', 'datasource'].includes(subject.type)
    )
      throw new Error('Subject must be a prisma generator or datasource!');

    function tap<T>(subject: T, callback: (s: T) => void) {
      callback(subject);
      return subject;
    }

    const assignment = subject.assignments.reduce<schema.Assignment>(
      (memo, assignment) =>
        assignment.type === 'assignment' && assignment.key === key
          ? tap(assignment, a => {
              a.value = `"${value}"`;
            })
          : memo,
      {
        type: 'assignment',
        key,
        value: `"${value}"`,
      }
    );

    if (!subject.assignments.includes(assignment))
      subject.assignments.push(assignment);

    return this;
  }

  private blockInsert(statement: schema.Break | schema.Comment): this {
    let subject = this.getSubject<schema.Block>();
    const allowed = ['datasource', 'enum', 'generator', 'model', 'view'];
    if (!subject || !('type' in subject) || !allowed.includes(subject.type)) {
      const parent = this.getParent<schema.Block>();
      if (!parent || !('type' in parent) || !allowed.includes(parent.type)) {
        throw new Error('Subject must be a prisma block!');
      }

      subject = this._subject = parent;
    }

    switch (subject.type) {
      case 'datasource': {
        subject.assignments.push(statement);
        break;
      }
      case 'enum': {
        subject.enumerators.push(statement);
        break;
      }
      case 'generator': {
        subject.assignments.push(statement);
        break;
      }
      case 'model': {
        subject.properties.push(statement);
        break;
      }
    }
    return this;
  }

  /** Add a line break */
  break(): this {
    const lineBreak: schema.Break = { type: 'break' };
    return this.blockInsert(lineBreak);
  }

  /**
   * Add a comment. Regular comments start with // and do not appear in the
   * prisma AST. Node comments start with /// and will appear in the AST,
   * affixed to the node that follows the comment.
   * */
  comment(text: string, node = false): this {
    const comment: schema.Comment = {
      type: 'comment',
      text: `//${node ? '/' : ''} ${text}`,
    };
    return this.blockInsert(comment);
  }

  /**
   * Add a comment to the schema. Regular comments start with // and do not appear in the
   * prisma AST. Node comments start with /// and will appear in the AST,
   * affixed to the node that follows the comment.
   * */
  schemaComment(text: string, node = false): this {
    const comment: schema.Comment = {
      type: 'comment',
      text: `//${node ? '/' : ''} ${text}`,
    };
    this.schema.list.push(comment);
    return this;
  }

  /**
   * Adds or updates a field in the current model. The field can be customized
   * further with one or more .attribute() calls.
   * */
  field(name: string, fieldType: string | schema.Func = 'String'): this {
    let subject = this.getSubject<schema.Object>();
    if (!isSchemaObject(subject)) {
      const parent = this.getParent<schema.Object>();
      if (!isSchemaObject(parent))
        throw new Error('Subject must be a prisma model or view!');

      subject = this._subject = parent;
    }

    const field = subject.properties.reduce<schema.Field>(
      (memo, block) =>
        block.type === 'field' && block.name === name ? block : memo,
      {
        type: 'field',
        name,
        fieldType,
      }
    );

    if (!subject.properties.includes(field)) subject.properties.push(field);
    this._parent = subject;
    this._subject = field;
    return this;
  }

  /** Drop a field from the current model or view. */
  removeField(name: string): this {
    let subject = this.getSubject<schema.Object>();
    if (!isSchemaObject(subject)) {
      const parent = this.getParent<schema.Object>();
      if (!isSchemaObject(parent))
        throw new Error('Subject must be a prisma model or view!');

      subject = this._subject = parent;
    }

    subject.properties = subject.properties.filter(
      field => !(field.type === 'field' && field.name === name)
    );
    return this;
  }

  /**
   * Returns the current subject, allowing for more advanced ways of
   * manipulating the schema.
   * */
  then<R extends schema.Block | schema.Property>(
    callback: (subject: R) => unknown
  ): this {
    callback(this._subject as R);
    return this;
  }
}

export function createPrismaSchemaBuilder(
  source?: string
): PrismaSchemaBuilder<
  Exclude<
    keyof ConcretePrismaSchemaBuilder,
    DatasourceOrGeneratorKeys | EnumKeys | FieldKeys | BlockKeys
  >
> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return new ConcretePrismaSchemaBuilder(source) as any;
}
