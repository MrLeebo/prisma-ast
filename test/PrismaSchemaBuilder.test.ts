import { createPrismaSchemaBuilder } from '../src/PrismaSchemaBuilder';
import * as schema from '../src/getSchema';
import { loadFixture } from './utils';

describe('PrismaSchemaBuilder', () => {
  it('adds a generator', () => {
    const builder = createPrismaSchemaBuilder();
    builder
      .generator('client', 'prisma-client-js')
      .assignment('output', 'client.js');
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      generator client {
        provider = \\"prisma-client-js\\"
        output   = \\"client.js\\"
      }
      "
    `);
  });

  it('updates an existing generator', () => {
    const builder = createPrismaSchemaBuilder();
    builder
      .generator('client', 'prisma-client-js')
      .assignment('output', 'client.js');
    builder.generator('client').assignment('engineType', 'library');
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      generator client {
        provider   = \\"prisma-client-js\\"
        output     = \\"client.js\\"
        engineType = \\"library\\"
      }
      "
    `);
  });

  it('accesses a generator', () => {
    const builder = createPrismaSchemaBuilder();
    builder.generator('client').then<schema.Generator>(generator => {
      const assignment = generator.assignments[0] as schema.Assignment;
      assignment.value = '"prisma-client-ts"';
    });
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      generator client {
        provider = \\"prisma-client-ts\\"
      }
      "
    `);
  });

  it('sets the datasource', () => {
    const builder = createPrismaSchemaBuilder();
    builder.datasource('postgresql', { env: 'DATABASE_URL' });
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      datasource db {
        url      = env(\\"DATABASE_URL\\")
        provider = postgresql
      }
      "
    `);
  });

  it('accesses the datasource', () => {
    const builder = createPrismaSchemaBuilder();
    builder
      .datasource('postgresql', { env: 'DATABASE_URL' })
      .then<schema.Datasource>(datasource => {
        datasource.name = 'my-database';
      });
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      datasource my-database {
        url      = env(\\"DATABASE_URL\\")
        provider = postgresql
      }
      "
    `);
  });

  it('can reset datasource url with assignments', () => {
    // not really sure why you'd do this, but *shrug*
    const builder = createPrismaSchemaBuilder();
    builder
      .datasource('postgresql', { env: 'DATABASE_URL' })
      .assignment('url', 'https://database.com');
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      datasource db {
        url      = \\"https://database.com\\"
        provider = postgresql
      }
      "
    `);
  });

  it('adds a model', () => {
    const builder = createPrismaSchemaBuilder();
    builder.model('Project').field('name', 'String');
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model Project {
        name String
      }
      "
    `);
  });

  it('removes a model', () => {
    const builder = createPrismaSchemaBuilder(`
      datasource db {
        url = env("DATABASE_URL")
      }

      model Project {
        name String
      }
    `);
    builder.drop('Project');
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      datasource db {
        url = env(\\"DATABASE_URL\\")
      }


      "
    `);
  });

  it('accesses a model', () => {
    const builder = createPrismaSchemaBuilder(`
    model Project {
      name String
    }
  `);
    builder.model('Project').then<schema.Model>(project => {
      project.name = 'Task';
    });
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model Task {
        name String
      }
      "
    `);
  });

  it('removes an enum', () => {
    const builder = createPrismaSchemaBuilder(`
    datasource db {
      url = env("DATABASE_URL")
    }

    enum Role {
      MEMBER
      ADMIN
    }
    `);

    builder.drop('Role');

    expect(builder.print()).toMatchInlineSnapshot(`
      "
      datasource db {
        url = env(\\"DATABASE_URL\\")
      }


      "
    `);
  });

  it('adds an enum', () => {
    const builder = createPrismaSchemaBuilder();
    builder
      .enum('Role', ['USER', 'ADMIN'])
      .break()
      .comment('test')
      .enumerator('OWNER');

    expect(builder.print()).toMatchInlineSnapshot(`
      "
      enum Role {
        USER
        ADMIN

        // test
        OWNER
      }
      "
    `);
  });

  it('updates an existing enum', () => {
    const builder = createPrismaSchemaBuilder();
    builder.enum('Role', ['USER', 'ADMIN']);
    builder.enum('Role').enumerator('OWNER');

    expect(builder.print()).toMatchInlineSnapshot(`
      "
      enum Role {
        USER
        ADMIN
        OWNER
      }
      "
    `);
  });

  it('accesses an enum', () => {
    const builder = createPrismaSchemaBuilder();
    builder.enum('Role', ['USER', 'ADMIN']).then<schema.Enum>(e => {
      e.name = 'UserType';
    });

    expect(builder.print()).toMatchInlineSnapshot(`
      "
      enum UserType {
        USER
        ADMIN
      }
      "
    `);
  });

  it('adds a field to an existing model', () => {
    const builder = createPrismaSchemaBuilder(`
    model Project {
      name String
    }
    `);
    builder.model('Project').field('description', 'String');
    builder.model('Project').field('owner', 'String');

    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model Project {
        name        String
        description String
        owner       String
      }
      "
    `);
  });

  it('updates an existing field', () => {
    const builder = createPrismaSchemaBuilder(`
    model TaskScript {
      name      String
      createdAt DateTime
      updatedAt DateTime
    }
    `);
    builder
      .model('TaskScript')
      .field('createdAt', 'DateTime')
      .attribute('default', [{ name: 'now' }]);
    builder
      .model('TaskScript')
      .field('updatedAt', 'DateTime')
      .attribute('updatedAt');
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model TaskScript {
        name      String
        createdAt DateTime @default(now())
        updatedAt DateTime @updatedAt
      }
      "
    `);
  });

  it('removes a field', () => {
    const builder = createPrismaSchemaBuilder(`
    model TaskScript {
      name      String
      createdAt DateTime
      updatedAt DateTime
    }
    `);
    builder
      .model('TaskScript')
      .removeField('createdAt')
      .removeField('updatedAt');
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model TaskScript {
        name String
      }
      "
    `);
  });

  it('adds an attribute', () => {
    const builder = createPrismaSchemaBuilder();
    builder
      .model('TaskMessage')
      .field('createdAt', 'DateTime?')
      .attribute('db.Timestamptz', ['6']);
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model TaskMessage {
        createdAt DateTime? @db.Timestamptz(6)
      }
      "
    `);
  });

  it('replaces an attribute', () => {
    const builder = createPrismaSchemaBuilder();
    builder
      .model('TaskMessage')
      .field('createdAt', 'DateTime?')
      .attribute('db.Timestamptz', ['6']);

    // Replace the @db.Timestamptz(6) attribute by dropping and re-creating the field
    builder
      .model('TaskMessage')
      .field('createdAt', 'DateTime?')
      .removeAttribute('db.Timestamptz')
      .attribute('default', [{ name: 'now' }]);
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model TaskMessage {
        createdAt DateTime? @default(now())
      }
      "
    `);
  });

  it('replaces an attribute with access', () => {
    // Set up the schema with a model and a field with an attribute
    const builder = createPrismaSchemaBuilder();
    builder
      .model('TaskMessage')
      .field('createdAt', 'DateTime?')
      .attribute('db.Timestamptz', ['6']);

    // Replace the @db.Timestamptz(6) attribute with @default(now())
    builder
      .model('TaskMessage')
      .field('createdAt')
      .then<schema.Field>(field => {
        const attribute: schema.Attribute = {
          kind: 'field',
          name: 'default',
          type: 'attribute',
          args: [{ type: 'attributeArgument', value: 'now()' }],
        };
        field.attributes = [attribute];
      });
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model TaskMessage {
        createdAt DateTime? @default(now())
      }
      "
    `);
  });

  it('adds a field relation', () => {
    const builder = createPrismaSchemaBuilder();
    builder
      .model('Project')
      .field('user', 'User')
      .attribute('relation', { fields: ['clientId'], references: ['id'] });
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model Project {
        user User @relation(fields: [clientId], references: [id])
      }
      "
    `);
  });

  it('adds a map attribute', () => {
    const builder = createPrismaSchemaBuilder();
    builder.model('Project').blockAttribute('map', 'projects');
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model Project {
        @@map(\\"projects\\")
      }
      "
    `);
  });

  it('adds an id attribute', () => {
    const builder = createPrismaSchemaBuilder();
    builder
      .model('User')
      .field('firstName', 'String')
      .field('lastName', 'String')
      .blockAttribute('id', ['firstName', 'lastName']);
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model User {
        firstName String
        lastName  String
        @@id([firstName, lastName])
      }
      "
    `);
  });

  it('adds a unique attribute', () => {
    const builder = createPrismaSchemaBuilder();
    builder
      .model('Project')
      .field('code', 'String')
      .field('client', 'String')
      .blockAttribute('unique', ['code', 'client']);
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model Project {
        code   String
        client String
        @@unique([code, client])
      }
      "
    `);
  });

  it('adds a comment', () => {
    const builder = createPrismaSchemaBuilder();
    builder
      .model('User')
      .comment('this is the part of the name you say first', true)
      .field('firstName', 'String')
      .comment('this is the part of the name you say last', true)
      .field('lastName', 'String')
      .break()
      .comment('this is a random comment')
      .break();

    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model User {
        /// this is the part of the name you say first
        firstName String
        /// this is the part of the name you say last
        lastName  String

        // this is a random comment

      }
      "
    `);
  });

  it('adds a schema comment', () => {
    const builder = createPrismaSchemaBuilder();
    builder
      .model('Project')
      .schemaComment('this is a comment')
      .schemaComment('this is a node comment', true)
      .model('Node');
    expect(builder.print()).toMatchInlineSnapshot(`
      "
      model Project {
        
      }
      // this is a comment
      /// this is a node comment

      model Node {
        
      }
      "
    `);
  });

  it('can reference the same attribute', () => {
    const builder = createPrismaSchemaBuilder(`
    model Test {
      id String @id @default(auto()) @map(\\"_id\\") @db.ObjectId
    }
    `);
    const result = builder
      .model('Test')
      .field('id', 'String')
      .attribute('id')
      .attribute('default', [{ name: 'auto' }])
      .attribute('map', [`"_id"`])
      .attribute('db.ObjectId')
      .print();
    expect(result).toMatchInlineSnapshot(`
    "
    model Test {
      id String @id @default(auto()) @map(\\"_id\\") @db.ObjectId
    }
    "
  `);
  });

  it('can create a view', () => {
    const builder = createPrismaSchemaBuilder(`
    model Project {
      name String
    }
    `);
    const result = builder
      .view('TestView')
      .field('id', 'String')
      .attribute('id')
      .attribute('default', [{ name: 'auto' }])
      .attribute('map', [`"_id"`])
      .attribute('db.ObjectId')
      .print();
    expect(result).toMatchInlineSnapshot(`
    "
    model Project {
      name String
    }

    view TestView {
      id String @id @default(auto()) @map(\\"_id\\") @db.ObjectId
    }
    "
  `);
  });

  it('edits an existing view', () => {
    const builder = createPrismaSchemaBuilder(`
    view TestView {
      id String
    }
    `);
    const result = builder
      .view('TestView')
      .field('name', 'String')
      .print();
    expect(result).toMatchInlineSnapshot(`
    "
    view TestView {
      id   String
      name String
    }
    "
  `);
  });

  it('prints the schema', async () => {
    const source = await loadFixture('example.prisma');
    const result = createPrismaSchemaBuilder(source).print();
    expect(result).toMatchInlineSnapshot(`
      "// https://www.prisma.io/docs/concepts/components/prisma-schema
      // added some fields to test keyword ambiguous

      datasource db {
        url      = env(\\"DATABASE_URL\\")
        provider = \\"postgresql\\"
      }

      generator client {
        provider = \\"prisma-client-js\\"
      }

      model User {
        id        Int       @id @default(autoincrement())
        createdAt DateTime  @default(now())
        email     String    @unique
        name      String?
        role      Role      @default(USER)
        posts     Post[]
        projects  Project[]
      }

      model Post {
        id         Int      @id @default(autoincrement())
        createdAt  DateTime @default(now())
        updatedAt  DateTime @updatedAt
        published  Boolean  @default(false)
        title      String   @db.VarChar(255)
        author     User?    @relation(fields: [authorId], references: [id])
        authorId   Int?
        // keyword test
        model      String
        generator  String
        datasource String
        enum       String // inline comment

        @@map(\\"posts\\")
      }

      model Project {
        client      User     @relation(fields: [clientId], references: [id])
        clientId    Int
        projectCode String
        dueDate     DateTime

        @@id([projectCode])
        @@unique([clientId, projectCode])
        @@index([dueDate])
      }

      model Model {
        id Int @id

        @@ignore
      }

      enum Role {
        USER // basic role
        ADMIN // more powerful role
      }

      model Indexed {
        id  String @id(map: \\"PK_indexed\\") @db.UniqueIdentifier
        foo String @db.UniqueIdentifier
        bar String @db.UniqueIdentifier

        @@index([foo, bar(sort: Desc)], map: \\"IX_indexed_indexedFoo\\")
      }
      "
    `);
  });
});
