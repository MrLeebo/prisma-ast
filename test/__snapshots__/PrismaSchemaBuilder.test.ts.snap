// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrismaSchemaBuilder adds a composite type 1`] = `
"
datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

model Product {
  id     String  @id @default(auto()) @map("_id") @db.ObjectId
  name   String
  photos Photo[]
}

type Photo {
  height Int    @default(0)
  width  Int    @default(0)
  url    String
}
"
`;
